project/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -28,25 +45,84 @@
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Extracts user input of some location.
+    String address = request.getParameter("address");
 
-    // Resolves that location to a particular political location unit.
+    // Resolve {@code address} to a list of election names.
+    // @TODO [Need to call the Civic Information API]
+    List<String> electionNames = Arrays.asList("New York's 14th Congressional District Election");
 
-    // Finds (1) official election/candidate information; perhaps a brief version.
-
-    // Sends data as response.
+    // Find election/candidate information. Package and convert the data to JSON.
+    List<Election> electionsData = findBriefElectionCandidateInformation(electionNames);
+    DirectoryPageDataPackage dataPackage = new DirectoryPageDataPackage(electionsData);
+    Gson gson = new Gson();
+    String dataPackageJson = gson.toJson(dataPackage);
 
+    // Send data in JSON format as the servlet response for the directory page.
+    response.setContentType("application/json;");
+    response.getWriter().println(dataPackageJson);
   }
 
   @Override
-  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.sendRedirect("/directory.html");
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {}
+
+  /**
+   * Queries the database for (brief version) official election/candidate information and format
+   * as {@code Election} objects.
+   */
+  private List<Election> findBriefElectionCandidateInformation(List<String> electionNames) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    // Query for (brief version) official election/candidate information, with election names.
+    List<Entity> elections = new ArrayList<>(electionNames.size());
+    for (String electionName : electionNames) {
+      Query electionQuery = new Query("Election")
+          .setFilter(new FilterPredicate("__key__", FilterOperator.EQUAL,
+                    KeyFactory.createKey("Election", electionName)));
+      PreparedQuery electionQueryResult = datastore.prepare(electionQuery);
+      elections.add(electionQueryResult.asSingleEntity());
```

Comment:
Instead of having 2 for loops where you first get the election entities and then populate electionsData, you could call a helper function that does what your second for loop does and call it something like extractCandidateInformation. That way you also just have one list electionsData

Reply:
Thanks for pointing this out! I've made the changes to `DataServlet`